kakao/src/main/kotlin/com/agoda/kakao/delegate/Delegate.kt

```diff
@@ -0,0 +1,57 @@
+package com.agoda.kakao.delegate
+
+import com.agoda.kakao.intercept.Interceptor
+
+/**
+ * Base delegate interface.
+ *
+ * Provides functionality of aggregating interceptors and invoking them on `check`
+ * and `perform` invocations.
+ *
+ * @see Interceptor
+ */
+interface Delegate<T, C, P> {
+    var interaction: T
+
+    fun viewInterceptor(): Interceptor<T, C, P>?
+    fun screenInterceptor(): Interceptor<T, C, P>?
+    fun kakaoInterceptor(): Interceptor<T, C, P>?
+
+    fun interceptors(): List<Interceptor<T, C, P>> = mutableListOf<Interceptor<T, C, P>>().also { list ->
+        viewInterceptor()?.let { list.add(it) }
+        screenInterceptor()?.let { list.add(it) }
+        kakaoInterceptor()?.let { list.add(it) }
+    }
+
+    fun interceptCheck(assertion: C): Boolean {
+        interceptors().forEach { interceptor ->
```

Comment:
Should we create a new list instance each time?

For something like
```
textViewLarge {
    isDisplayed()
    hasText("text")
    click()
}
```
a list of interceptors will be created 3 times. Maybe we could cash it and update if intercept() is called on the kview level?

Reply:
I suppose the answer is 'yes'.
But it gives a guarantee that at any moment we have an actual list of interceptors and we don't afraid to change a configuration of interceptors.

Reply:
But the commit is valid, it's a waste of memory. I will refactor it to use fixed array, keeping the memory usage low and still guaranteeing the valid set of interceptors at each invocation.